clients/src/main/java/org/apache/kafka/common/requests/DeleteAclsRequest.java

```diff
@@ -34,107 +36,84 @@
-import static org.apache.kafka.common.protocol.CommonFields.RESOURCE_PATTERN_TYPE_FILTER;
-import static org.apache.kafka.common.protocol.CommonFields.RESOURCE_TYPE;
 
 public class DeleteAclsRequest extends AbstractRequest {
-    private final static String FILTERS = "filters";
-
-    private static final Schema DELETE_ACLS_REQUEST_V0 = new Schema(
-            new Field(FILTERS, new ArrayOf(new Schema(
-                    RESOURCE_TYPE,
-                    RESOURCE_NAME_FILTER,
-                    PRINCIPAL_FILTER,
-                    HOST_FILTER,
-                    OPERATION,
-                    PERMISSION_TYPE))));
-
-    /**
-     * V1 sees a new `RESOURCE_PATTERN_TYPE_FILTER` that controls how the filter handles different resource pattern types.
-     * For more info, see {@link PatternType}.
-     *
-     * Also, when the quota is violated, brokers will respond to a version 1 or later request before throttling.
-     */
-    private static final Schema DELETE_ACLS_REQUEST_V1 = new Schema(
-            new Field(FILTERS, new ArrayOf(new Schema(
-                    RESOURCE_TYPE,
-                    RESOURCE_NAME_FILTER,
-                    RESOURCE_PATTERN_TYPE_FILTER,
-                    PRINCIPAL_FILTER,
-                    HOST_FILTER,
-                    OPERATION,
-                    PERMISSION_TYPE))));
-
-    public static Schema[] schemaVersions() {
-        return new Schema[]{DELETE_ACLS_REQUEST_V0, DELETE_ACLS_REQUEST_V1};
-    }
-
     public static class Builder extends AbstractRequest.Builder<DeleteAclsRequest> {
-        private final List<AclBindingFilter> filters;
+        private final DeleteAclsRequestData data;
 
-        public Builder(List<AclBindingFilter> filters) {
+        public Builder(DeleteAclsRequestData data) {
             super(DELETE_ACLS);
-            this.filters = filters;
+            this.data = data;
+        }
+
+        public Builder(List<AclBindingFilter> filters) {
+            this(new DeleteAclsRequestData().setFilters(filters.stream()
+                    .map(DeleteAclsRequest::deleteAclsFilter).collect(Collectors.toList())));
         }
 
         @Override
         public DeleteAclsRequest build(short version) {
-            return new DeleteAclsRequest(version, filters);
+            validate(version);
+            return new DeleteAclsRequest(version, data);
         }
 
         @Override
         public String toString() {
-            return "(type=DeleteAclsRequest, filters=" + Utils.join(filters, ", ") + ")";
+            return "(type=DeleteAclsRequest, filters=" + Utils.join(data.filters(), ", ") + ")";
+        }
+
+        private void validate(short version) {
+            if (version == 0) {
+                for (DeleteAclsRequestData.DeleteAclsFilter filter : data.filters()) {
+                    PatternType patternType = PatternType.fromCode(filter.patternTypeFilter());
+
+                    // If ANY is specified, we override it to the default of LITERAL
```

Comment:
There should be a little more explanation about why we're doing this, I think.  The explanation would be something like this: on older brokers, no pattern types existed except LITERAL (effectively...).  So even though ANY is not directly supported on those brokers, we can get the same effect as ANY by setting LITERAL on those particular brokers.

Reply:
Also, it's really ugly for "validate" to mutate the request.  Can we have another function for this... :pray:

Reply:
@cmccabe Older brokers supported LITERAL and `*`, right?

Reply:
"*" is considered "LITERAL" for compatibility reasons